common/argparse.h

```diff
@@ -0,0 +1,895 @@
+  // command whose arguments are being parsed and to do the parsing. Everything
+  // is done in a single invocation as the common case is to build a command
+  // description, parse the arguments once, and then run with that
+  // configuration.
+  //
+  // The `out` stream is treated like `stdout` would be for a Unix-style command
+  // line tool, and `errors` like `stderr`: any errors or diagnostic information
+  // are printed to `errors`, but meta-actions like printing a command's help go
+  // to `out`.
+  static auto Parse(llvm::ArrayRef<llvm::StringRef> unparsed_args,
+                    llvm::raw_ostream& out, llvm::raw_ostream& errors,
+                    const CommandInfo& command_info,
+                    llvm::function_ref<void(CommandBuilder&)> build)
+      -> ParseResult;
+
+ private:
+  class Parser;
+
+  friend auto operator<<(llvm::raw_ostream& output, ParseResult result)
+      -> llvm::raw_ostream& {
+    switch (result) {
+      case ParseResult::Error:
+        return output << "Error";
+      case ParseResult::MetaSuccess:
+        return output << "MetaSuccess";
+      case ParseResult::Success:
+        return output << "Success";
+    }
+  }
+
+  friend auto operator<<(llvm::raw_ostream& output, ArgKind kind)
+      -> llvm::raw_ostream& {
+    switch (kind) {
+      case ArgKind::Flag:
+        return output << "Boolean";
+      case ArgKind::Integer:
+        return output << "Integer";
+      case ArgKind::String:
+        return output << "String";
+      case ArgKind::OneOf:
+        return output << "OneOf";
+      case ArgKind::MetaActionOnly:
+        return output << "MetaActionOnly";
+      case ArgKind::Invalid:
+        return output << "Invalid";
+    }
+    CARBON_FATAL() << "Corrupt argument kind!";
+  }
+
+  friend auto operator<<(llvm::raw_ostream& output, CommandKind kind)
+      -> llvm::raw_ostream& {
+    switch (kind) {
+      case CommandKind::Invalid:
+        return output << "Invalid";
+      case CommandKind::RequiresSubcommand:
+        return output << "RequiresSubcommand";
+      case CommandKind::Action:
+        return output << "Action";
+      case CommandKind::MetaAction:
+        return output << "MetaAction";
+    }
+    CARBON_FATAL() << "Corrupt command kind!";
+  }
+};
+
+// The internal representation of a parsable argument description.
+struct Args::Arg {
+  using Kind = ArgKind;
+  using ValueActionT =
+      std::function<bool(const Arg& arg, llvm::StringRef value_string)>;
+  using DefaultActionT = std::function<void(const Arg& arg)>;
+
+  explicit Arg(const ArgInfo& info);
+  ~Arg();
+
+  ArgInfo info;
+  Kind kind = Kind::Invalid;
+  bool has_default = false;
+  bool is_required = false;
+  bool is_append = false;
+  bool is_help_hidden = false;
+
+  // Meta action storage, only populated if this argument causes a meta action.
+  ActionT meta_action = {};
+
+  // Storage options depending on the kind.
+  union {
```

Comment:
Why a union instead of a variant? It seems like a variant could be used because I wouldn't expect this to be performance sensitive. The advantage of a variant is it simplifies the destructor.

(I'm hoping I'm not accidentally repeating this comment)